host_core/lib/host_core/policy/manager.ex

```diff
@@ -0,0 +1,160 @@
+        action: action,
+        host: %{
+          public_key: HostCore.Host.host_key(),
+          lattice_id: HostCore.Host.lattice_prefix(),
+          labels: HostCore.Host.host_labels(),
+          valid_cluster_issuers: HostCore.Host.cluster_issuers()
+        }
+      }
+      |> evaluate(topic)
+      |> cache_decision(source, target, action)
+    else
+      :policy_eval_disabled ->
+        allowed_action("Policy evaluation disabled, allowing action", "")
+
+      {:ok, decision} ->
+        decision
+
+      {:error, invalid_error} ->
+        Logger.error("#{invalid_error}")
+
+        allowed_action(invalid_error)
+    end
+  end
+
+  defp evaluate(req, topic) do
+    case Jason.encode(req) do
+      {:ok, encoded} ->
+        case HostCore.Nats.safe_req(:control_nats, topic, encoded,
+               receive_timeout: HostCore.Policy.Manager.policy_timeout()
+             ) do
+          {:ok, %{body: body}} ->
+            # Decode body with existing atom keys
+            case Jason.decode(body, keys: :atoms!) do
+              {:ok, policy_res} ->
+                policy_res
+
+              {:error, _decode} ->
+                allowed_action(
+                  "Policy response failed to decode, allowing",
+                  req |> Map.get(:request_id, "not supplied")
+                )
+            end
+
+          {:error, :timeout} ->
+            allowed_action(
+              "Policy request timed out, allowing action",
+              req |> Map.get(:request_id, "not supplied")
+            )
+        end
+
+      {:error, e} ->
+        Logger.error("Could not JSON encode request, #{e}")
+
+        allowed_action("", req |> Map.get(:request_id, "not supplied"))
+    end
+  end
+
+  # Returns nil if not present or {:ok, decision} based on previous policy decision
+  defp cached_decision(source, target, action) do
+    case :ets.lookup(@policy_table, {source, target, action}) do
+      [{{_src, _tgt, _act}, decision}] -> {:ok, decision}
+      [] -> nil
+    end
+  end
+
+  # Inserts a policy decision into the policy table as a nested tuple. This
+  # allows future lookups to easily fetch decision based on {source,target,action}
+  defp cache_decision(decision, source, target, action) do
+    :ets.insert(@policy_table, {{source, target, action}, decision})
+    decision
+  end
+
+  ##
+  # Basic validation of source, target, and action ensuring required fields are present
+  ##
+  defp validate_source(%{
+         public_key: _public_key,
+         capabilities: _caps,
+         issuer: _issuer,
+         issued_on: _issued_on,
+         expires_in_mins: _expires
+       }) do
+    :ok
+  end
+
+  defp validate_source(source) when is_map(source) do
+    # Narrow down missing fields by removing present fields from the list
+    missing_fields =
+      [:public_key, :capabilities, :issuer, :issued_on, :expires_in_mins]
+      |> Enum.reject(fn required_field -> Map.get(source, required_field) != nil end)
```

Comment:
`reject` with `!= nil` hurts my brain. Why not `filter` with `== nil`?

Reply:
good call!